packages/inter-protocol/src/auction/auctionBook.js

```diff
@@ -0,0 +1,319 @@
+        currencyAvailable,
+        curAuctionPrice,
+      );
+      collateralRecord = {
+        Collateral: affordableCollateral,
+      };
+      currencyRecord = {
+        Currency: currencyAvailable,
+      };
+    }
+
+    trace('settle', { currencyRecord, collateralRecord });
+
+    atomicRearrange(
+      zcf,
+      harden([
+        [collateralSeat, seat, collateralRecord],
+        [seat, currencySeat, currencyRecord],
+      ]),
+    );
+    return collateralRecord.Collateral;
+  };
+
+  const isActive = auctionState => auctionState === AuctionState.ACTIVE;
+
+  const acceptOffer = (seat, price, want, timestamp, auctionState) => {
+    trace('acceptPrice');
+    // Offer has ZcfSeat, offerArgs (w/price) and timeStamp
+
+    let collateralSold = AmountMath.makeEmptyFromAmount(want);
+    if (isActive(auctionState) && ratioGTE(price, curAuctionPrice)) {
+      collateralSold = settle(seat, want);
+
+      if (AmountMath.isEmpty(seat.getCurrentAllocation().Currency)) {
+        seat.exit();
+        return;
+      }
+    }
+
+    const stillWant = AmountMath.subtract(want, collateralSold);
+    if (!AmountMath.isEmpty(stillWant)) {
+      priceBook.add(seat, price, stillWant, timestamp);
+    } else {
+      seat.exit();
+    }
+  };
+
+  const acceptDiscountOffer = (
+    seat,
+    discount,
+    want,
+    timestamp,
+    auctionState,
+  ) => {
+    trace('accept discount');
+    let collateralSold = AmountMath.makeEmptyFromAmount(want);
+
+    if (
+      isActive(auctionState) &&
+      isDiscountedPriceHigher(discount, curAuctionPrice, lockedPrice)
+    ) {
+      collateralSold = settle(seat, want);
+      if (AmountMath.isEmpty(seat.getCurrentAllocation().Currency)) {
+        seat.exit();
+        return;
+      }
+    }
+
+    const stillWant = AmountMath.subtract(want, collateralSold);
+    if (!AmountMath.isEmpty(stillWant)) {
+      discountBook.add(seat, discount, stillWant, timestamp);
+    } else {
+      seat.exit();
+    }
+  };
+
+  return Far('AuctionBook', {
+    addAssets(assetAmount, sourceSeat) {
+      trace('add assets');
+      assetsForSale = AmountMath.add(assetsForSale, assetAmount);
+      atomicRearrange(
+        zcf,
+        harden([[sourceSeat, collateralSeat, { Collateral: assetAmount }]]),
+      );
+    },
+    settleAtNewRate(reduction) {
+      curAuctionPrice = multiplyRatios(reduction, lockedPrice);
+
+      const pricedOffers = priceBook.offersAbove(curAuctionPrice);
+      const discOffers = discountBook.offersAbove(reduction);
+
+      // requested price or discount gives no priority beyond specifying which
+      // round the order will be service in.
+      const prioritizedOffers = [...pricedOffers, ...discOffers].sort(
+        ([a], [b]) => TimeMath.compareAbs(keyToTime(a), keyToTime(b)),
+      );
+
+      trace(`settling`, pricedOffers.length, discOffers.length);
+      prioritizedOffers.forEach(([key, { seat, price: p, wanted }]) => {
```

Comment:
prefer syntax to function calls.
```suggestion
      for (const ([key, { seat, price: p, wanted }]) of prioritizedOffers) {
```

Reply:
why is this better?  To me, it looks roughly equivalent.

Reply:
I also prefer syntactic `for` loops over `forEach` specifically. (For the other higher order array methods, I have no general preference, but evaluate case by case.)

Two reasons I prefer syntax:
   * usually easier to step through iterations in a debugger, depending on the debugger
   * easy to exit early, which often arises after the code is initially written

Reply:
done